documentation/memory.md

```diff
@@ -0,0 +1,89 @@
+# Modular memory
+
+The design of the DINO CPU modular memory system encourages abstraction of memory functionality into distinct memory components in a robust, flexible manner.
+
+## Basic components
+
+The modular memory system utilizes a 'pipeline', 'memory ports', and a 'backing memory':
+
+##### Pipeline
+The pipeline is simply the CPU model issuing memory operations. From the perspective of the pipeline, the memory is composed of `MemPortIO` interfaces only, 
+
+##### Memory ports
+Memory ports are smart modules which transform incoming pipeline signals in a `MemPortIO` into a `Request`. Likewise, it receives `Response`s from the backing memory and pre-processes the response data to result in a typical `readdata` output.
+
+The memory ports abstract the gap between the pipeline and the backing memory in a way that permits flexibility: while a port is fundamentally responsible for exchanging information between the pipeline and the backing memory, the method by which it does this can vary with specialized implementations of memory ports like caches.
+
+##### Backing memory
+The backing memory is a module which wraps the inner `Mem` object and performs writes or reads to the inner memory depending on the request it receives.
+
+Like the memory ports, the backing memory abstraction permits specialized implementations which handle memory requests immediately before executing them. For example, latency is achieved in the non-combinational memories by inserting all `Request`s into a `Pipe` before they are processed.
+
+Backing memories see the ports and pipeline as being composed only of `MemPortBusIO` interfaces.
+
+## Stable interfaces
+Each component connects to each other using two interfaces: `MemPortIO` and `MemPortBusIO`. 
+
+These memory interfaces are stable between all CPU models, so all memory components should communicate using these 
+interfaces _only_. 
+
+
+##### MemPortIO
+`MemPortIO` is the primary abstract interface for memory ports. As such, it has IO corresponding with the signals a port receives from the pipeline, and a `Flipped[MemPortBusIO]` which connects to a backing memory.
+
+Inputs: 
+- address: The address of the location in memory
```

Comment:
I suggest writing:

```
- `address`: The address of the location in memory
```

It may also be a good idea to copy the IO into the documentation here.

Also, is there any assumptions on the alignment of the address?

Reply:
This is simply addresses being aligned to the 32-bit block widths, correct?

Reply:
If there's right, sure. I don't remember what your design was :).